src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
I don't follow the logic, I don't think we will expose more than 1 constructor for system grains, so selection would always be trivial. Also, when using the DefaultGrainActivator, the selection is what we typically use. It's when using a different GrainActivator, including ServeBasedGrainActivator, that users will be able to specify how the grain is resolved, and hence can change the default. Of course if they don't override anything for system grains, they will be resolved as expected, so this more flexible activation doesn't go against how system grains are activated.
If people move away from defaults, they could of course shoot themselves in the foot, as with any extensibility feature, but I'd argue they would have to go way out of their way to mess the constructor selection part for system grains.
Again, I don't understand why constructor selection *needs* to be well defined. Can you explain the reason for it? because this is where we start deviating in the rest of the details. ASP.NET in particular does not put this constraint, and I don't understand why we should be different and deviate from that, and be a lot more restrictive than them. I don't see any advantage to it.
Also, what is the purpose of providing a different IGrainActivator if object activation is mostly fixed?

> Constructor selection needs be a well defined behavior because it must be the same for internal Orleans grains, and grains used in extensions like the providers. This is not a new constraint introduced by this PR.

This IS a new constraint for this PR. Where was that constraint enforced before? Sure, for providers we currently don't have an extensibility point to override the usage of ActivatorUtilities, so it currently only picks the same constructor every time, regarless of the container. But this is absolutely not a constraint today for grains, when defining your own IGrainActivator.

And to reiterate: what is the need to couple it to GrainTypeData in this PR? this was just for an optimization to not do reflection every time. This optimization does not warrant creating a new constraint, as it is absolutely not harmless, and there are trivial ways to avoid the coupling, as already stated.